Zerobase - MVC

Kova700 opened the pull request:
HomeFragment에서 아이템에 있는 하트를 누르고,
스크롤 아래로 내린 후 다시 위로 돌아올 때,
ListAdapter내부 currentList에는 Heart값이 해당 아이템이 그대로 false인 채로  남아 있어서
다시 뷰홀더에 데이터가 binding될 떄, 하트가 그대로 false로 남아 있는 현상이 발생합니다.
PokemonListItem 클래스의 Heart 프로퍼티 값을 수정가능하게 var로 변경해서
하트가 눌리면 homeAdapter.currentList에 있는 아이템의 heart값을 수정해주면 이런 현상을 방지할 수 있습니다.

하지만
불변 값(val)으로 data class를 구성하는 게 데이터 전달 시에 안정적이라는 것을 알기에
불변 값으로 data class를 구성하고 매번 버튼을 누를 때마다, submitList를 해주는 게 맞는건지 고민입니다.

Kova700 committed:
[feat] home,wish fragment 파일 추가

Kova700 committed:
[config] compileSdk버전 33 -> 34로 수정

Kova700 committed:
[feat(activity_main.xml)] BottomNavigation 추가

1. 바텀 네비게이션 layout 추가
2. 관련 icon 리소스 추가
3. 바텀 네비 menu 추가

Kova700 committed:
[feat(MainActivity.kt)] Fragment 화면 전환 트랜젝션 처리

1. Fragment전환시 스크롤 유지를 위해서 show & hide 방식으로 구성

Kova700 committed:
[refector] Fragment Transaction commitNow로 수정

1. Fragment BackStack을 쓰지 않고, 동기적으로 안정적인 Fragment 접근을위해서

Kova700 committed:
[config(build.gradle)] Glide의존성 추가 및 기타 의존성 버전업

1. absoluteAdapterPosition호출이 안되길래 recyclerview 최신버전 의존성 추가
2. core-ktx의존성 버전업

Kova700 committed:
[config(Manifest.xml)] Internet user permission 선언 추가

Kova700 committed:
[feat] HomeFrament.xml과 관련 리소스 추가

1. 하트.svg파일 + selector 추가

Kova700 committed:
[style(activity_main.xml)] xml파일 컨벤션 정리

Kova700 committed:
[style(MainActivity.kt)] 바텀네비 lateinit초기화 방식으로 수정

1. 기존에 lazy by구조로 초기화를 헀었지만, Fragment와 통일성을 갖고자 수정

Kova700 committed:
[feat] HomeFragment RecyclerView로직 초안 추가

1. HomeAdapter.kt에서 ItemClick 이벤트, 데이터 동기화 로직 제외 완성
2. 더미 데이터 클래스 추가 (임시)
3. Fragment간 데이터 동기화 로직이 필요함

Kova700 committed:
[style] HomeFragment의 RecyclerView로직 공용 클래스로 변경

1. WishFragment와 HomeFragment가 사용하는 데이터가 같음으로

Kova700 committed:
[style(HomeFragment.kt)] apply함수 적용,requireActivity()수정

Kova700 committed:
[style(R.layout.item_pokemon_list)]이름 변경(이전 : R.layout.item_home)

Kova700 committed:
[중간점검]

theo-f-lab commented on app/src/main/java/com/kova700/zerotomvvm/SampleDummy.kt:
```diff
@@ -0,0 +1,20 @@
+fun getDummy(): MutableList<PokemonListItem> {
+    val dummyList = mutableListOf<PokemonListItem>()
+
+    for (i in 1..30) {
```
```
(1..30).forEach { i -> 
}
```

theo-f-lab commented on app/src/main/java/com/kova700/zerotomvvm/MainActivity.kt:
```diff
@@ -1,11 +1,98 @@
+import com.google.android.material.bottomnavigation.BottomNavigationView
 
 class MainActivity : AppCompatActivity() {
+    private lateinit var bottomNavigationView: BottomNavigationView
```
```
private val bottomNavigationView by lazy {
   findViewById<BottomNavigationView>(R.id.bottomNavigationView)
}
```

Kova700 committed:
[fix(wish,homeFragement)] 부모 Activity onStart에서 초기화되게 수정

1. 매번 부모 액티비티 접근할 때 캐스팅해주던 코드를 onStart에서 미리
   초기화

Kova700 committed:
[style] onCreateView에서 layout.xml연결 코드 생성자로 대체

1. 생성자에 layout xml 전달함으로써 oncreateVeiw 삭제

Kova700 committed:
[feat(DetailActivity.kt)] detatil Activity 추가 + 화면 전환 기능 추가

Kova700 committed:
[fix] show,hide방식으로 fragment 버그 해결

1. home 탭과 wish 탭에서 동기화 문제 해결
2. replace + msavedViewState방식으로 수정해보자

Kova700 committed:
[refector] fragemnt replace방식으로 수정

0. 알게된 점
1. replace방식으로 그릴 시에 화면이 다시 그려짐
2. 생명주기를 이용하기가 hide/show방식보다 용이함
3. 같은 fragment 객체를 이용할 시 repace방식도 스크롤이 유지된다.
4. fragment 객체 내부적으로 스크롤 포지션이 저장되고 그 위치로 다시 복왼되는 구조
5. show/hide 방식은 그냥 아무 작업이 없음으로 화면이 유지되는 구조

Kova700 committed:
[feat(DetailActivity.kt)] detailActivty UI + 데이터 송수신 구현

Kova700 committed:
[style(PokemonItemClickListener.kt)] 불필요한 주석 삭제

Kova700 committed:
[feat] 버튼 누르면 랜덤데이터 추가 기능

theo-f-lab commented on app/src/main/AndroidManifest.xml:
```diff
@@ -14,6 +14,9 @@
         tools:targetApi="31">
+        <activity
+            android:name=".DetailActivity"
+            android:exported="false" />
```
질) `android:exported="false"` 부분이 어떤 의미일까요? + 기본값이 뭘까요?

Kova700 replied:
다른 앱 혹은 시스템 컴포넌트에서 해당 Activity를 실행할 수 있는지 유무를 나타내는 옵션입니다.

[공식문서](https://developer.android.com/topic/security/risks/android-exported?hl=ko)를 참고하여 보니 기본값은 
API 레벨 16(Android 4.1.1) 이하에는 true로 되어있지만, 버전에 따라 기본값이 다름으로, 
명시적으로 true와 false 둘 중 하나를 꼭 지정해주는 걸 권장하고 있습니다.

이 옵션을 사용하는 예로는 
1. 딥링크를 이용할 때, 특정 URL에 접속 시 시스템이 지정된 Activity를 띄울 수 있게 허용할 때, 해당 옵션을 사용할 수 있습니다.
2. LAUNCHER 카테고리의 intent-filter를 가진 Activity는 앱 시작 시에 시스템이 해당 App의 Activity를 실행할 수 있어야함으로 android:exported 옵션은 항상 true이어야합니다.

theo-f-lab commented on app/src/main/java/com/kova700/zerotomvvm/DetailActivity.kt:
```diff
@@ -0,0 +1,63 @@
+    private val pokemonListItem by lazy {
+        intent.getSerializableExtraData(TO_DETAIL_SELECTED_ITEM_EXTRA, PokemonListItem::class.java)
+            ?: throw Exception("pokemon is not exist")
+    }
```
질문) 조금더 적절한 예외처리는 어떤게 있을까요? 실제 사용할때 excepiton이 발생하면 사용을 하지 못하게될건데, 적절한 예외처리가 어떤게 더 좋을까요? 한번 고민해보시면 좋을것 같아요.

Kova700 replied:
DetailActivity로 데이터를 전달하게 구현을 해놨지만, 만약 예상치 못하게 excepiton이 발생한다면, 
DetailActivity에는 노출할 수 있는 정보가 없을 뿐더러, 
예외처리가 되어있지 않음으로 App자체가 강제종료되는 상황이 생길 수 있습니다.

고로, 넘겨받은 값이 없다면 Empty 객체 값을 넘겨주고, 
CustomException을 정의해서 ErrorHandle을 해주거나, 엘비스 연산자 뒤에 
따로 예외처리 함수를 호출해주는 구조를 가지는 것이 좋을 것 같습니다.

예외 처리 내용으로는 Toast 혹은 SnackBar 혹은 Dialog를 이용해서
유저에게 정상적으로 작동하지 않는 상황임을 인지시켜주는 것이 좋을 것 같습니다.

theo-f-lab replied:
근데 이 경우는 예외처리가 필요 없죠. 어차피 데이터가 없으면 비정상적인 접근이니.
어차피 기본 적으로 제공되어지는 입셉션으로 충분합니다.

사실 불필요함을 알려드릴려고 이야기한거에요.
진행에 꼭 필요한 데이터인데. 굳이 입셉션을 낼 필요도 없을거고, 기본적인 형태로도 충분하구요.

Kova700 replied:
아!... 그렇군요,,, 

"!!" 사용을 무의식적으로 계속 피하다보니 
매번 throw를 던지는 식으로 구성했던 것같습니다.

이해헀습니다! 감사합니다!

theo-f-lab commented on app/src/main/java/com/kova700/zerotomvvm/DetailActivity.kt:
```diff
@@ -0,0 +1,63 @@
+        heartToggleButton.apply {
+            isChecked = pokemonListItem.heart
+            setOnClickListener {
+                pokemonListItem.heart = !pokemonListItem.heart
```
`!`도 좋은데 `.not()`을 사용하는것도 가능해요.

theo-f-lab commented on app/src/main/java/com/kova700/zerotomvvm/DetailActivity.kt:
```diff
@@ -0,0 +1,63 @@
+        onBackPressedDispatcher.addCallback {
+            finishActivity()
+        }
+    }
```
질문) 이 코드의 용도는 뭔가요?

Kova700 replied:
유저가 뒤로가기 버튼을 누를 때, 작동할 callback을 정의 할 수 있습니다.
따로 UI상에 뒤로가기 버튼을 만들지 않았기 떄문에,
DetailActivity에서 Back 버튼을 누를 시에 , 
유저가 변경한 값을 intent에  ExtraData로 전달하기 위해서 추가한 코드입니다.

theo-f-lab replied:
음 onDestory에서도 충분히 동일한 행위를 할 수 있지 않나요?
어차피 이렇게 종료하나 저렇게 종료하나 종료 케이스는 많을건데 백버튼 말구도요.

그리고 사용자 액션상 취소가 아닐지?

Kova700 replied:
그렇군요,,
어쩌다 보니 자연스레 버튼 눌러서 종료, 백버튼 눌러서 종료인 경우만 있다고 생각했었습니다.
말씀해주신 내용대로 onDestroy로 이전해서 코드 반영했습니다!!

Kova700 replied:
결과를 넘겨주는  Activity에서 registerForActivityResult의 반환 값을 넘겨줄 때,
해당 Activity의 onDestroy가 호출되기 전에, 이전 Activity로 돌아와버려서,

해당 Activity의 onDestroy부분에 setReuslt를 지정하는 방식을 통해서는
결과 값이 registerForActivityResult에 정의해놓은 콜백에 넘어오지 않는 걸 확인했습니다.

마땅히 다른 괜찮은 방법이 떠오르지 않아서 다음 PR까지 반영해서 올리도록 하겠습니다!

theo-f-lab commented on app/src/main/java/com/kova700/zerotomvvm/DetailActivity.kt:
```diff
@@ -0,0 +1,63 @@
+        heartToggleButton.apply {
+            isChecked = pokemonListItem.heart
+            setOnClickListener {
+                pokemonListItem.heart = !pokemonListItem.heart
```
`!`을 쓰면 간혹 못보는 경우가 있는데 이럴땐 `.not()`을 쓰면 눈에 더 잘보이긴해요. 참고하세요.

Kova700 replied:
Wow...
not()이란 함수가 있었군요,,
감사합니다 😇

theo-f-lab commented on app/src/main/java/com/kova700/zerotomvvm/DetailActivity.kt:
```diff
@@ -0,0 +1,63 @@
+    private fun finishActivity() {
+        intent.putExtra(TO_MAIN_HEART_BOOLEAN_EXTRA, pokemonListItem.heart)
+        intent.putExtra(TO_MAIN_ITEM_POSITION_EXTRA, intent.getIntExtraData(TO_DETAIL_ITEM_POSITION_EXTRA))
+        finishWithExtra()
```
Intent를 새로 생성해서 finish하는것과, 지금처럼 기존 intent를 처리하는게 어떤 차이가 있을까요?

Kova700 replied:
Activity 내부에서 Intent 객체를 새로 생성하지 않고, intent 프로퍼티로 접근할 떄의 intent는 
해당 Activity를 startActivity로 호출했을 떄, 사용된 Intent를 뜻합니다.

현재 Activity에서 다른 Activity를 호출하는 경우가 아니라면,
불필요하게 Intent를 생성하지 않고 기존의 Intent를 이용해서도
반환 데이터를 전달 할 수 있음으로 기존 Intent를 통해 데이터를 전달하는 구조를 취했습니다.

새로운 Activity를 호출해야되는 상황에도 Activity의 intent 프로퍼티를 사용할 수 있지만,
이전 Activity에서 저장한 데이터도 같이 전달되기 때문에,
불필요한 데이터를 전달하기보단 새로운 Intent 객체를 생성해서 이용하는 것이 좋다고 생각합니다.

theo-f-lab replied:
코드상 보시면 같은 걸 다시 그대로 사용하는 코드가 보여지네요.
어차피 새로 만드실거라면 Intent()하는편이 더 맞을것 같죠.

Kova700 replied:
반영했습니다!!

theo-f-lab commented on app/src/main/java/com/kova700/zerotomvvm/HomeFragment.kt:
```diff
@@ -1,59 +1,91 @@
+
+    override fun onStart() {
+        super.onStart()
+        mainActivity = requireActivity() as? MainActivity ?: throw Exception("Unknown Activity")
```
질문) fragment 라이프사이클 중 어디에서 처리하는게 더 좋을까요? onStart는 언제 동작할까요?

Kova700 replied:
onAttach( )가 가장 먼저 Fragment lifecycle에서 호출되고,
Fragment가 inflate된 Activity의 Context를 이미 가지고 있습니다.
그렇기 떄문에, 전역변수 초기화의 목적을 위해서라면 미리 앞에서 초기화하는 방식으로
onStart( )보단 onAttach( )에서 mainActivity 변수를 초기화하는 것이 가독성 측면에서는 
더 괜찮은 방법이라고 생각합니다.

onStart( )는 
onViewCreated( )와 onResume( ) 사이에 호출되는데,
사용자와 상호작용은 되지 않지만, Fragment가 사용자에게 보여질 수 있을 떄 호출됩니다. 

Activity의 onStart( )와 마찬가지로 빠르게 onResume( )으로 넘어가기 때문에,
onStart( )에 작성할 코드가 있다면 
onResume( )(사용자와 상호작용 가능)나 onViewCreated( )에서 처리하는 방향으로 구현을 하면 됨으로,
굳이 코드를 작성할 일이 별로 없는 생명주기 콜백입니다.

Kova700 replied:
추가로, onViewCreated에서 `mainActivity`에 접근하고 있는데, 
onStart()에서 mainActivity를 초기화 하고 있었네요,,,

theo-f-lab replied:
이론적인 부분을 아는것 같은데... 코드에 반영이 안되어지는것 같네요.

Kova700 replied:
수정사항 반영했습니다!!

theo-f-lab commented on app/src/main/java/com/kova700/zerotomvvm/HomeFragment.kt:
```diff
@@ -1,59 +1,91 @@
+                    mainActivity.homePokeymonList[itemPosition].copy(heart = heartValue)
+                homeAdapter.submitList(mainActivity.homePokeymonList.toList())
+            }
+    }
```
registerForActivityResult는 언제 생성해두면 될까요? 매번 생성해둘 필요가 있을까요?

Kova700 replied:
Activity에서는 registerForActivityResult를 멤버 변수에서 호출하여 인스턴스 생성 시에만 호출되게 하는 것이 좋아보이나,

Fragment의 경우에는 어떤 식으로 Fragment를 전환하는 지에 따라, 객체의 재사용 여부 등 Framgment 사용 방법에 따라, 다른 방법을 채택하는 것이 좋아보입니다. 

만약 제 코드처럼 Fragment를 replace방식으로 전환하고, 인스턴스를 재활용한다면 
onDestroy()콜백을 탈 때마다, registerForActivityResult에 등록했던 콜백이 unregister됨으로
onAttach() 혹은 onCreate()에서 매번 등록을 해주는 방향으로 구현했습니다.

이에 관해서 헷갈리는 내용이 많아서 따로 [블로그](https://duft-doft.tistory.com/58)에 정리를 해봤습니다.

Kova700 replied:
Activity에서는 onResume()이전 상태에서 호출해두면 됨으로,
registerForActivityResult를 멤버 변수에서 호출하여 인스턴스 생성 시에만 호출되게 하는 것이 좋아보입니다.

Fragment의 경우에서도 onViewRestore() 이전 상태에서 호출해두면 됨으로,
registerForActivityResult를 멤버 변수에서 호출하여 인스턴스 생성 시에만 호출되게 하는 것이 좋을 것이라고 생각했었습니다. 

하지만, 만약 제 코드처럼 Fragment를 replace방식으로 전환하고, 인스턴스를 재사용한다면
replace될 때마다, onDestroy() 콜백이 호출됨으로, registerForActivityResult에 등록했던 콜백이 unregister된다는 것을 확인 했습니다. 그래서 onAttach() 혹은 onCreate()에서 매번 등록을 해주는 방향으로 구현했습니다.

이에 관해서 헷갈리는 내용이 많아서 따로 [블로그](https://duft-doft.tistory.com/58)에 정리를 해봤습니다.

Kova700 replied:
fragment를 재사용하는 것이 [안티 패턴](https://issuetracker.google.com/issues/247221861)이라는 것을 뒤늦게 알게 되어서,
이전 Fragment의 상태(스크롤 등)를 새로운 Fragment에 주입 시키는 방식으로 수정해서
매번 새로운 Fragment를 사용하는 방향으로 수정해보겠습니다!

theo-f-lab replied:
initActivityResultLauncher 을 한번만 생성하나요? N번 생성하나요? 지금 올려주신 내용이랑 코드가 다른 부분은 없나요?

theo-f-lab commented on app/src/main/java/com/kova700/zerotomvvm/HomeFragment.kt:
```diff
@@ -1,59 +1,91 @@
+
+                        //하지만
+                        //불변 값(val)으로 data class를 구성하는 게 데이터 전달시에 안정적이라는 것을 알기에
+                        //불면 값으로 data class를 구성하고 매번 버튼을 누를 때마다, submitList를 해주는 게 맞는건지 고민입니다.
```
이 부분은 DiffUtil의 역할을 충분히 이해하면 될것 같은데.
data class를 왜 불변으로 둬야하는지도요.

Kova700 replied:
하나의 아이템의 상태가 변경될 때마다
기존에 Adapter가 가지고 있던 많은 수의 아이템들을 복사해서 
새로운 List를 다시 생성해, 그 List를 Adapter에게 submitList 한다면, 

매번 아이템의 상태를 변경할 때마다
ItemSize가 몇 일지도 모르는 List를 생성해야 하고,
그 List를 처음부터 다시 검사하여 기존 List와의 차이점을 찾는다는 것이

List 비교 후 데이터 차이가 없다면 UI가 바뀌지 않는다고 하더라도,
완전히 데이터 set이 바뀐 것도 아니고, 하나의 아이템만 바뀐 것을 알고 있는 상황에서
매번 List를 생성하고, 처음부터 다시 읽는다는 것이 비효율적이지 않나라는 생각을 했었습니다.

말씀해 주신 내용대로,
DiffUtil과 ListAdapter에 대해서 좀 더 알아 본 결과

ListAdapter에 submitList를 하면, 
내부적으로 AsyncListDiffer를 이용해 backgroundThread가 넘겨받은 리스트와 기존 리스트의 차이점을 확인하고,
메인 Thread는 그 차이점만 UI 변화에 반영한다는 걸 인지하지 못하고 있었고,

ListAdapter의 사용 목적 중 하나가,
notifyItemChanged(), notifyItemInserted() 등의 notifyItem 함수들의 번거로운 사용과
안정적인 UI 업데이트를 위해서 사용한다는 것을 다시 인지했습니다.

고로,, List를 새로 다시 만드는 것이 효율적인 측면에서 보기 싫다면,
RecyclerViewAdapter를 사용해서 notifyItem 함수들을 효율적이게 사용하면 되지만,
실수할 확률이 높아지고, 코드가 번잡해진다는 점을 감안해야 함도 알게 되었습니다.
🙇‍♂️

theo-f-lab replied:
음? 마지막 부분이 어떤 의미일까요?

데이터의 아이템 하나하나를 비교하는것이 아니라, 리스트 자체를 비교하고, 데이터가 다르면 다음 아이템을 비교하겠죠? 이보다 더 효율적인 방법을 어떻게 만드실 수 있다는건지 이해를 못하겠네요.

> 고로,, List를 새로 다시 만드는 것이 효율적인 측면에서 보기 싫다면,
RecyclerViewAdapter를 사용해서 notifyItem 함수들을 효율적이게 사용하면 되지만,
실수할 확률이 높아지고, 코드가 번잡해진다는 점을 감안해야 함도 알게 되었습니다.

theo-f-lab replied:
그리고 데이터가 변하지 않는데 DiffUtil을 왜 탄다는건지 역시 이해를 못하겠네요.
리스트 안에 데이터가 변했으니 DiffUtil이 동작을 할것인데 어떤 비효율이 생기나요?

theo-f-lab commented on app/src/main/java/com/kova700/zerotomvvm/MainActivity.kt:
```diff
@@ -1,98 +1,53 @@
+        supportFragmentManager.beginTransaction()
+            .setReorderingAllowed(true)
+            .replace(R.id.container_main, targetFragment)
+            .commitNow()
```
질문) show/hide로 처리하던것과 어떠한 차이가 생기나요?

Kova700 replied:
우선,
## hide & show 방식으로 Fragment전환 시에는 
Fragment의 생명 주기가 onResume()까지만 호출된 채로, 
기존의 Fragment를 Destroy하지 않고, 
hide되었던 Fragment를 다시 show하는 방식으로 전환이 이루어집니다.

기존의 인스턴스가 살아 있기 때문에, 
Fragment 내부의 RecyclerView의 scroll위치 등 chiledView들의 state를 잃지 않습니다.
hide & show 방식이 매번 Fragment 전환 시마다 
Fragment 인스턴스를 새로 생성하지 않아도 되는 장점이 있기도 하지만,

Fragment 전환 시에 더 이상의 생명 주기 변화가 없음으로
만약 Fragment 생명 주기 별로 콜백에 특정한 작업을 하고 싶다면 
repalce를 이용해 Fragment 전환을 이용하는 것이 좋습니다.

(hide & show 방식으로 Fragment가 hide 혹은  show 될 때마다 특정 트리거가 꼭 필요하다면, 
Custom으로 LifecycleOwner를 만들어서 구현이 가능하긴 합니다. [참고](https://pluu.github.io/blog/android/2023/01/19/fragment_visible_lifecycleowner/))

## replace 방식으로 Fragment전환 시에는 
기존에 inflate되어 있던 Fragment를 Destroy하고
새로운 Fragment를 inflate합니다.

hide & show 방식과 달리 매번 새로운 인스턴스를 생성해서 사용하기 때문에,
스크롤 포지션을 포함한 ChideView들의 state를 보존하지 못합니다.

hide & show 방식과 달리 Fragment 전환 시마다 Fragment의 생명주기 콜백을 처음부터 다시 타기 떄문에,
각 생명주기 별로 특정한 작업을 할 일이 있는 경우에는 replace방식이 hide & show 방식보다 더 적합한 방식입니다.

하지만, replace방식을 사용해도 Fragment 내부에 있는 RecyclerView의 스크롤이 유지되는 상황이 좀 있는데

1. Configuration Change가 발생할 때,
Activity의 onSaveInstanceState를 통해 Fragment의 state가 Activity의 state에 저장됨으로,
onRestoreInstanceState에서 새로 생성되는 Activity에게 state를 주입합니다.
그러므로 Fragment의 state가 유지됩니다. 
(Activity도 새로운 객체, Fragment도 새로운 객체지만, state가 주입되는 상황)
그래서 스크롤이 유지되는 것처럼 보입니다. 
(따로 개발자가 정의하지 않아도 supertype에서 알아서 작업해줍니다.)

2. Destroy된 Fragment instance를 재사용하는 경우
Fragment가 Destroy되기 전에 Fragment 내부에 state를 저장하는데, 
해당 state를 가지고 있기 때문에 destroy 되었던 Fragment로 replace하면
스크롤 위치가 그대로 보존됩니다.

하지만 Fragment 재사용은 [안티 패턴](https://issuetracker.google.com/issues/247221861)이라고 하니,
Fragment의 state를 Destroy되기 전에 가져와서, 새로운 Fragment의 argument로 주입하는 방식으로 구현하는 게
좋은 방법이라고 생각됩니다.

theo-f-lab commented on app/src/main/java/com/kova700/zerotomvvm/WishFragment.kt:
```diff
@@ -1,52 +1,73 @@
-        container: ViewGroup?,
-        savedInstanceState: Bundle?
-    ): View? = inflater.inflate(R.layout.fragment_wish, container, false)
+    private lateinit var mainActivity: MainActivity
```
질문) `mainActivity`를 이렇게 직접 들고있는데, 만약 `WishFragmnet`가 DetailActivity에서도 사용되어야 한다면 어떤방식으로 처리하는것이 좋을까요?

Kova700 replied:
mainActivity 변수를 MainActivity의 상위 타입인 AppCompatActivity타입으로 정의하고,
mainActivity 변수로 MainActivity 클래스 내부에 정의된 함수들의 사용하는 경우,
mainActivity의 타입이 MainActivity인지 확인하는 로직을 추가해서
사용하는 방식으로 처리하면 될 것 같습니다!

theo-f-lab replied:
왜 굳이 Activity를 사용하려고 하시는걸지... Interface를 사용하시면 매우 쉬운

theo-f-lab replied:
```
interface OnActivity {
    fun getList(): List<...>
    val list: List<...>
}
```

```
class MainActivity : OnActivity {}

class DetailActivity : OnActivity {}
```

```
class WishFragment {

    private var onActivity: OnActivity? = null
    override fun onAttach(context: Context) {
          onActivity = (context as? OnActivity)
    }
}
```

theo-f-lab commented:
다음에 ViewBinding 적용해보세요.

theo-f-lab commented on app/src/main/java/com/kova700/zerotomvvm/HomeFragment.kt:
```diff
@@ -36,7 +38,13 @@ class HomeFragment : Fragment(R.layout.fragment_home) {
     private fun View.connectViewComponent() {
-        recyclerview = this@connectViewComponent.findViewById<RecyclerView>(R.id.rcv_home_fragment)
+        recyclerview = this@connectViewComponent.findViewById(R.id.rcv_home_fragment)
+        itemAddBtn = this@connectViewComponent.findViewById(R.id.fab_home_fragment)
```
이미 this scope이 connectViewComponent인데 this@를 중복으로 정의하실거면... 그냥 파라메터로 받는편이 더 좋겠죠.
